BJS2-40651 Project Service : Vacancies - Create Vacancy Feature

drewbal opened the pull request.

drewbal committed:
Initial commit

drewbal committed:
BJS2-40651 Feature: Create Vacancy - adds VacancyDto

DTO: VacancyDto added

drewbal committed:
BJS2-40651 Feature: Create Vacancy - adds Vacancy Mapper

DTO: VacancyDto optimized
Mapper: VacancyMapper added

drewbal committed:
BJS2-40651 Feature: Create Vacancy - adds controller and create method

VacancyController: createVacancy method added

drewbal committed:
BJS2-40651 Feature: Create Vacancy - create full vacancy from dto

VacancyService: toEntityFromDto method added (add fields ignored by mapper)
ProjectService: getProjectById metod added

drewbal committed:
BJS2-40651 Feature: Create Vacancy - validate data to create the vacancy

VacancyValidator: checks the project is valid; role of the creator is valid;
TeamMemberService: returns the team member by id
DataValidationException: new Exception

drewbal committed:
BJS2-40651 Feature: Create Vacancy - creates new vacancy in system

VacancyService: creates a new vacancy;

YTXIRE commented on src/main/java/faang/school/projectservice/controller/VacancyController.java:
```diff
@@ -0,0 +1,26 @@
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("api/vacancies")
```
Добавь / перед api

drewbal replied:
Готово. Учту на будущее

YTXIRE commented on src/main/java/faang/school/projectservice/controller/VacancyController.java:
```diff
@@ -0,0 +1,26 @@
+        return ResponseEntity.status(HttpStatus.CREATED).body(vacancyService.create(vacancyDto));
+    }
+
+
```
Убери лишние пробелы

YTXIRE commented on src/main/java/faang/school/projectservice/service/VacancyService.java:
```diff
@@ -0,0 +1,32 @@
+        return vacancyMapper.toDto(vacancy);
+    }
+
+    public Vacancy toEntityFromDto(VacancyDto dto) {
+        Vacancy vacancy = vacancyMapper.toEntity(dto);
+        vacancy.setProject(projectService.getProjectById(dto.getProjectId()));
+        return vacancy;
+    }
```
А почему этот метод публичный?

drewbal replied:
Хороший пойнт. У него чисто утилитарная функция, поэтому можно спрятать, согласен

YTXIRE commented on src/main/java/faang/school/projectservice/validator/VacancyValidator.java:
```diff
@@ -0,0 +1,34 @@
+package faang.school.projectservice.validator;
```
Давай класть каждую сущность в подходящий пакет. А именно валидотор Vacancy в пакет vacancy. Чтобы если  что - то еще добавится связанное с ним, то было в отдельном пакете.

drewbal committed:
BJS2-40651 Feature: Create Vacancy - sets status at creation

VacancyDto: VacationStatus field added;
VacancyService: sets OPEN status at creation;

YTXIRE commented on src/test/java/faang/school/projectservice/service/ProjectServiceTest.java:
```diff
@@ -0,0 +1,73 @@
+                .id(1L)
+                .name("Project 1")
+                .build();
+
```
Тоже самое, убери лишние enter.

YTXIRE commented on src/test/java/faang/school/projectservice/service/ProjectServiceTest.java:
```diff
@@ -0,0 +1,73 @@
+                .build();
+
+    }
+
```
Тоже самое, убери лишние enter.

YTXIRE commented on src/test/java/faang/school/projectservice/service/TeamMemberServiceTest.java:
```diff
@@ -0,0 +1,56 @@
+                .userId(1L)
+                .roles(List.of(TeamRole.OWNER))
+                .build();
+
```
Тоже самое, убери лишние enter.

YTXIRE commented on src/test/java/faang/school/projectservice/service/VacancyServiceTest.java:
```diff
@@ -0,0 +1,124 @@
+    @Test
+    @DisplayName("Create a new vacancy successfully")
+    void createSuccess() {
+        when(vacancyMapper.toEntity(dto)).thenReturn(vacancy);
```
Не проще ли использовать обычный маппер?

drewbal replied:
@Spy был в прошлой задаче, в этой хочу попробовать @Mock и почувствовать разницу

drewbal committed:
BJS2-40651 Feature: Create Vacancy - review fixes

VacancyService: toEntityFromDto is private;
VacancyController: path fixed;
Other: blank lines removed

DeemaP commented on src/main/java/faang/school/projectservice/service/ProjectService.java:
```diff
@@ -0,0 +1,20 @@
+public class ProjectService {
+    private final ProjectRepository projectRepository;
+
+    public Project getProjectById(Long projectId) {
```
Если подтянуть ProjectJpaRepository, то у тебя будет прямой метод у репозитория findById, который возвращает Optional<Project>, заодно и валидировать проще)

drewbal replied:
Ты не поверишь, но в ProjectRepository уже есть такой метод с проверкой :)

<img width="892" alt="image" src="https://github.com/user-attachments/assets/a6526895-fce0-4f0e-8561-96bf91160653">

DeemaP replied:
Справедливо) Тут гребаная путаница началась с этими репозиториями)

DeemaP commented on src/main/java/faang/school/projectservice/service/ProjectService.java:
```diff
@@ -0,0 +1,20 @@
+        return projectRepository.getProjectById(projectId);
+    }
+
+    public boolean checkProjectExistsById(Long projectId) {
```
А зачем этот метод, да еще и в сервисе? У тебя же и так projectRepository.existsById(projectId) возвращает boolean.

drewbal replied:
Тут у нас эта непонятная путаница с репозиториями. Я использовал ProjectRepository и включенные в него методы. Но ты тоже прав, одно и то же можно получить разными методами разных репо.

Зачем мне такой странный метод. Он мне нужен в валидаторе вакансий, а тащить в него репозиторий моветон. Поэтому метод через сервис

DeemaP replied:
В смысле тащить репозиторий в валидатор моветон?)
У нас в юзер сервисе все валидаторы в себя тянут свои репозитории)

drewbal replied:
Ну не у всех :) у меня все общение с чужими репозиториями построено через их сервисе.
Пока у меня в голове следующая парадигма: любая бизнес-логика и направление запросов в БД происходит через сервисы. А валидаторы сфокусированы только на проверке логики, им в бд самостоятельно не надо лезть. По этому же принципу мы не делаем всю валидацию на стороне контроллера, например. То есть использование сервисов повышает инкапсуляцию системы и ее компонентов. А мы стремимся к принципу единой ответственности (S), но при этом кроме проверки данных даем валидаторам функцию работы с базой.

Во-вторых, мне кажется, что упрощение связей зависимостей повышает устойчивость. Т.е. мы убираем  прямую зависимость валидатора от изменений в репозитории.

DeemaP commented on src/main/java/faang/school/projectservice/validator/vacancy/VacancyValidator.java:
```diff
@@ -0,0 +1,34 @@
+    private final TeamMemberService teamMemberService;
+
+    public void validateProjectInVacancyExists(VacancyDto dto) {
+        if (!projectService.checkProjectExistsById(dto.getProjectId())) {
```
А, вот эта валидация куда убежала)) 

Не странновато выглядит, что ты в валидатор притягиваешь валидирующий метод из сервиса?

Думаю тут два варианта есть:
1) Затащить прям сюда projectRepository и на нем проверять existsById
2) Сделать  projectValidator, который уже работает напрямую со своим репозиторием и из него вытянуть эту проверку)

У меня уже сделан ProjectValidator, но пока без этого метода) Потом смержим в один красивый)

drewbal replied:
Неть, не странновато. 
1) Я валидирую бизнес-логику вакансий, которая опирается на бд по проектам. Я не уверен, что тащить в валидатор репозиторий хорошая идея. Доступ к бд через репозиторий должны иметь только сервисы. Тем более тащить чужой репозиторий, ведь валидатор у меня для сервиса вакансий.
2) Валидатор в валидатор? Не думал над таким. Пока в моей картине четкое разграничение функций, то есть валидатор функционально обеспечивает проверку бизнес-логики своего сервиса. Т.е. projectValidator будет трудиться на благо сущности Project. Но мы можем подумать и порисовать.

DeemaP replied:
Погоди, ты в методе валидируешь не бизнес-логику, а просто наличие проекта в репозитории и все.
Поэтому логично положить его в ProjectValidator.

у меня в MentorshipRequestService был приватный метод, который назывался validateAndGetMentorshipRequest, вот он как бы и валидатор был, но и возвращал значение, поэтому я его сделал приватным в сервисе. А тут у тебя чистая валидация


![Screenshot 2024-11-07 at 09 27 58](https://github.com/user-attachments/assets/32056ec8-6127-4b49-92e7-369a451b8be2)

drewbal replied:
Я понимаю твою логику, и у меня несколько другой взгляд (не обязательно правильный). Я сторонник более гранулированной структуры проекта, где у каждой сущности своя инфраструктура (контроллер - сервис - валидатор - репозиторий). Движение в сторону централизации и обобщения с одной стороны упрощает систему, но с другой стороны работает против ее устойчивости из-за сильной зависимости. 

Смотри, ведь если еще дальше продвинуть твою логику, то окажется, что все компоненты так или иначе крутятся вокруг project (или user на прошлой неделе). Тогда всю валидацию можно поместить в projectValidator. Да и контроллер сделать один с путем `/api/v1/application/`, а в нем все вызовы обрабатывать. Да и единого сервиса-МФЦ хватит :) но это сделает проект монолитом, и любое изменение будет затрагивать все.

Я понимаю и принимаю твою позицию, но пока не могу осознать истинный путь (которого скорее всего нет и зависит от принятых в команде стандартов). Касательно твоего кода выше, пока у меня в голове мысль, что сервис не должен валидировать, поэтому методу validateWhatever место в валидаторе.

Но наверное классно, что мы озадачиваемся и дискутируем вопросы некоего системного дизайна, пусть и примитивного уровне

DeemaP replied:
Естественно ты не обязан менять свой код из-за моих комментариев!

Мы дискутируем, и у каждого своя точка зрения.

Но ведь ты сам говоришь validateWhatever - место в валидаторе, так почему у тебя метод checkProjectExistsById лежит в сервисе? Я в своем коде сделал приватный метод, который использует метод валидатора, а потом возвращает значение из репозитория)

А у тебя выходит чистая валидация, просто ты назвал ее не validate а check.

DeemaP replied:
НО! У меня в коде косяк и лишние строки, ведь вместо валидации на существование запроса я мог просто сделать обработку Optional от findById.

 Вот как полезно дискутировать))

drewbal replied:
Я понял тебя. Да, ты прав. У меня просто очень простая валидация и вытягивание метода из сервиса в валидатор для проверки этого примитивного критерия выглядит довольно странно. Так, а если мы захотим сделать более сложную проверку? Что вакансия существует и ее дату создания или количество кандидатов? Это же будет в валидаторе, но при этом включать в себя так же и проверку на наличие других критериев? Вот я все, что так или иначе касается валидации отношу в валидатор. То есть сервис в данном случае просто проверяет наличие в базе, но не делает вывод на что это может влиять.
Да, ситуация :)

DeemaP replied:
Упрощая весь диалог, процитирую тебя - "Валидатору - валидаторово".
Кто-то потом захочет тоже проверить есть ли проект в базе и будет искать этот метод в валидаторе, а он у тебя в сервисе лежит)

drewbal replied:
Да!! Именно так, "Валидатору - валидаторово"! Но валидатор сам не лезет в БД, а делает это через сервис, получает инфу, и делает вывод по ней. 

И если кто-то захочет проверить проект, как ты правильно говоришь, то за проверкой он пойдет в ТВОЙ сервис, в котором должен лежать этот метод. В моем окне ответственны за вакансии, а не за проекты. 

То есть я поручаю своему валидатору сделать запрос в другой репо. Он делает это через соответствующий сервис и делает проверку для собственного сервиса.

DeemaP commented on src/test/java/faang/school/projectservice/controller/VacancyControllerTest.java:
```diff
@@ -0,0 +1,69 @@
+        assertEquals("Vacancy 1", resultDto.getName());
+    }
+
+    private VacancyDto createTestVacancyDto() {
```
А это так надо? Не проще в beforeEach оставить? Или ты чисто чтоб глаза не мозолило?

drewbal replied:
Я бы сказал, что так можно, а не обязательно надо :) просто подумалось, что BeforeEach будет перегружен сильно, поэтому спустил служебные методы в подвал. Целиком дело вкуса

DeemaP reviewed (commented):
Намана делаешь) Но я тебе накинул на вентилятор, жду ответ))

drewbal committed:
BJS2-40651 Feature: Create Vacancy - review fixes

VacancyValidator: logic changed, VacancyRepository injected
VacancyService: validation fix
ProjectValidator: basic validation added

drewbal committed:
Merge branch 'Dragon-master-stream7' into BJS2-40651-create-vacancy

# Conflicts:
#	src/main/java/faang/school/projectservice/exception/DataValidationException.java
#	src/main/java/faang/school/projectservice/service/TeamMemberService.java

drewbal committed:
BJS2-40651 Feature: Create Vacancy - merge and conflicts solving

Minor fixes

drewbal committed:
BJS2-40651 Feature: Create Vacancy: review fixes

VacancyController: logging added
ServiceController: logging added
GlobalExceptionHandler: fixed wrong import, logging added
Other: spaces fixed

drewbal committed:
BJS2-40651 Feature: Create Vacancy: optimization

NewVacancyDto: new DTO for vacancy creation
VacancyController: createVacancy method accepts new dto with minimum required fields
VacancyMapper: takes new vacancy DTO, returns default DTO
VacancyService: create method optimized for new dto
VacancyValidator: role validation methods united

RuslanZinovyev commented on src/main/java/faang/school/projectservice/dto/vacancy/NewVacancyDto.java:
```diff
@@ -0,0 +1,48 @@
+    @Size(min = 10, max = 600, message = "Vacancy description must be at least 10 and under 600 characters long")
+    private String description;
+
+    @NotNull(message = "Project id must be a positive integer")
```
Сообщение очевидно должно быть другим.

drewbal replied:
Я придерживался идеи, что пользователю говорим в обоих случаях то, что нужно от него (положительное целое), а не говорим про его ошибку (прислал null). Но вообще согласен. 
Исправлено

RuslanZinovyev commented on src/main/java/faang/school/projectservice/dto/vacancy/NewVacancyDto.java:
```diff
@@ -0,0 +1,48 @@
+    @Positive(message = "Project id must be a positive integer")
+    private long projectId;
+
+    @NotNull(message = "Created by id cannot be empty")
```
`createdBy` довольно странное имя переменной для типа Long. Может лучше тогда `createdById`?

RuslanZinovyev commented on src/main/java/faang/school/projectservice/dto/vacancy/NewVacancyDto.java:
```diff
@@ -0,0 +1,48 @@
+    @Positive(message = "Created by id must be a positive integer")
+    private Long createdBy;
+
+    @NotNull(message = "Slavery is prohibited")
```
Ну, за чувство юмора хвалю.

RuslanZinovyev commented on src/main/java/faang/school/projectservice/dto/vacancy/NewVacancyDto.java:
```diff
@@ -0,0 +1,48 @@
+
+    @NotNull(message = "Slavery is prohibited")
+    @Positive(message = "Salary must be a positive number")
+    private Double salary;
```
Я думаю что на данном этапе это не так критично, но на будущее когда дело касается финансов всегда используй BigDecimal

drewbal replied:
Согласен. Пока я использовал тот формат, который уже был в сущности в системе. Я пока не знаю как эта переменная взаимодействует в проекте в целом, поэтому пока не трогаю. Но поизучаю вопрос

RuslanZinovyev commented on src/main/java/faang/school/projectservice/dto/vacancy/NewVacancyDto.java:
```diff
@@ -0,0 +1,48 @@
+    private Integer count;
+
+    @NotNull(message = "Required skills ids list cannot be empty")
+    @Size(min = 1, message = "Required skills ids list cannot be empty")
```
Можно 2 аннотации заменить на одну, она проверяет и на NULL и на пустой лист. 

`@NotEmpty(message = "Required skills ids list cannot be null or empty")`

The NotEmpty annotation makes use of the NotNull class, and also checks that the size/length of the supplied object is greater than zero.

drewbal replied:
Действительно!

RuslanZinovyev commented on src/main/java/faang/school/projectservice/dto/vacancy/VacancyDto.java:
```diff
@@ -0,0 +1,54 @@
+
+    @NotNull(message = "Project id must be a positive integer")
+    @Positive(message = "Project id must be a positive integer")
+    private long projectId;
```
Используешь примитив, хотя раннее использовал объект Long. Нужно придерживаться одной стратегии.
Потом очевидно ты по ошибке скопировал одинакое сообщение для обеих аннотаций.

drewbal replied:
Да, вижу, что null тут лишний.

Данная ДТО сделана исключительно как формат для отправки данных пользователю, поэтому лучше обзову ее ResponseVacancyDto, и поля приведу к соответствующим типам, предусматривающим пустые значения, если они отсутствуют в системе по какой-то причине. Например, createdById будет отсутствовать у вновь созданной вакансии. 
И я уберу аннотации валидации с нее, т.к. она не будет приходить извне, только отправляться

RuslanZinovyev commented on src/main/java/faang/school/projectservice/dto/vacancy/VacancyDto.java:
```diff
@@ -0,0 +1,54 @@
+
+    @NotNull(message = "Required skills ids list cannot be empty")
+    @Size(min = 1, message = "Required skills ids list cannot be empty")
+    private List<Long> requiredSkillIds;
```
Также можно заменить на NotEmpty

drewbal replied:
Вообще убрал валидацию для исходящего дто

RuslanZinovyev commented on src/main/java/faang/school/projectservice/service/ProjectService.java:
```diff
@@ -0,0 +1,20 @@
+@Service
+@RequiredArgsConstructor
+public class ProjectService {
+    private final ProjectRepository projectRepository;
```
Тут никакой валидации не должно быть на методах? Ты уверен что null услоаный сюда не прилетит?

drewbal replied:
Тут не нужна валидация в данном случае. Созданы всего 2 утилитарных метода, которые достают данные из бд через репозиторий, и там уже зашиты проверки на null с выбросом исключений

RuslanZinovyev commented on src/test/java/faang/school/projectservice/controller/VacancyControllerTest.java:
```diff
@@ -0,0 +1,85 @@
+
+    @Test
+    @DisplayName("Create a new vacancy successfully")
+    void testCreateVacancySuccess() {
```
Неплохо было бы также проверить и негативные сценарии, что будет если метод выбросит исключение и как поведет себя твое приложение в этом случае. Что будет если DTO не будет содержать верных параметров, можно и валидацию потестить.

drewbal replied:
Про негативный сценарий в целом согласен. И потестил в нем сценарий, если на этапе валидации через контроллер пролетает исключение

RuslanZinovyev commented on src/test/java/faang/school/projectservice/service/ProjectServiceTest.java:
```diff
@@ -0,0 +1,71 @@
+import static org.mockito.Mockito.*;
+
+@ExtendWith(MockitoExtension.class)
+class ProjectServiceTest {
```
Также рекомендую потестить негативные сценарии.

drewbal committed:
BJS2-40651 Feature: Create Vacancy: review fixes

Vacancy: createdBy -> createdById, updatedBy ->updatedById
NewVacancyDto: validation annotations fixed
VacancyDto: renamed in VacancyResponseDto to improve logic
Other: exception messages fixed
